api/blast_ui.api.inc

```diff
@@ -1,6 +1,102 @@
 <?php
 
 /**
+ * This program is used for services module specifically.
+ * Run BLAST (should be called from the command-line)
+ *
+ * @param $program
+ *   Which BLAST program to run (ie: 'blastn', 'tblastn', tblastx', 'blastp','blastx')
+ * @param $query
+ *   The full path and filename of the query FASTA file
+ * @param $database
+ *   The full path and filename prefix (excluding .nhr, .nin, .nsq, etc.)
+ * @param $output_filestub
+ *   The filename (not including path) to give the results. Should not include file type suffix
+ * @param $options
+ *   An array of additional option where the key is the name of the option used by
+ *   BLAST (ie: 'num_alignments') and the value is relates to this particular
+ *   BLAST job (ie: 250)
+ * @param $format 
+ *   Which specifies Json or asn format of file or program
+ * @param  $format_type 
+ * which specify the number of format used by BLAST e.g json has 11 and json as 12. 
+ */
+
+
+function run_BLAST_tripal_job_services($program, $query, $database, $output_filestub, $options, $format, $format_type,  $job_id = NULL) {
+
+
+  
+  $output_file = file_directory_temp() .  DIRECTORY_SEPARATOR . $output_filestub . '.blast.' . $format_type;
+
+  
+ // $output_file = file_directory_temp() .  DIRECTORY_SEPARATOR . $output_filestub . '.blast.json';
+  
+  $output_dir = variable_get('file_public_path', conf_path() . '/files') 
+              . DIRECTORY_SEPARATOR . 'tripal' . DIRECTORY_SEPARATOR . 'tripal_blast';
+
+  print "\nExecuting $program\n\n";
+  print "Query: $query\n";
+  print "Database: $database\n";
+  print "Results File: $output_file\n";
+
+  print "Options:\n";
+
+  // Allow administrators to use an absolute path for these commands.
+  // Defaults to using $PATH.
+	$blast_path    = variable_get('blast_path', '');
+	$blast_threads = variable_get('blast_threads', '');
+	
+	// Strip the extension off the BLAST target
+	$database = preg_replace("/(.*)\.[pn]\w\w/", '$1', $database);
+	
+	// The executables:
+	$program       = $blast_path . $program;
+	$blast_formatter_command = $blast_path . 'blast_formatter';
+
+  //$blast_cmd = "$program -query '$query' -db '$database' -out '$output_file' -outfmt=11";
+  $blast_cmd = "$program -query '$query' -db '$database' -out '$output_file' -outfmt=".$format;
```

Comment:
We have to be very careful at this point to ensure we don't bring in a security risk. When generating the blast command we have to make sure to escapeshellarg() every single parameter as we add it to the command. This ensures that if our validation fails at some point prior we don't end up with arbitrary code executed. See https://github.com/tripal/tripal_blast/blob/7.x-1.x/api/blast_ui.api.inc#L215 for an example.